### PR TITLE
[std]: Use XXHash32 instead FNV-1a

### PR DESCRIPTION
**Pull request by MaxGraey**
XXHash32 has better quality and speed (especially for 32-bit, 64-bit and >= 16 bytes keys).

**FNV-1a**:

- Poor uniform distribution. Only up to 14 bits can be used for the hash table (up to 2 ^ 14 hash table slots) without degradation
- Bad avalanche effect
- Speed: 4.5 cyclers per byte

**XXHash**:
- Excellent uniform distribution
- Near-perfect avalanche effect
- Speed: ~1.7 cyclers per byte

[string perf tests](https://www.assemblyscript.org/editor.html#IyFvcHRpbWl6ZT1zcGVlZCZydW50aW1lPXN0dWImbm9Bc3NlcnQKaW1wb3J0IHsgSEFTSCB9IGZyb20gInV0aWwvaGFzaCI7CgpAaW5saW5lIGNvbnN0IFhYSDMyX1AxOiB1MzIgPSAyNjU0NDM1NzYxOwpAaW5saW5lIGNvbnN0IFhYSDMyX1AyOiB1MzIgPSAyMjQ2ODIyNTE5OwpAaW5saW5lIGNvbnN0IFhYSDMyX1AzOiB1MzIgPSAzMjY2NDg5OTE3OwpAaW5saW5lIGNvbnN0IFhYSDMyX1A0OiB1MzIgPSA2NjgyNjUyNjM7CkBpbmxpbmUgY29uc3QgWFhIMzJfUDU6IHUzMiA9IDM3NDc2MTM5MzsKQGlubGluZSBjb25zdCBYWEgzMl9TRUVEOiB1MzIgPSAwOwoKCi8vIEB0cy1pZ25vcmU6IGRlY29yYXRvcgpAaW5saW5lCmZ1bmN0aW9uIG1peChoOiB1MzIsIGtleTogdTMyKTogdTMyIHsKICByZXR1cm4gcm90bChoICsga2V5ICogWFhIMzJfUDIsIDEzKSAqIFhYSDMyX1AxOwp9CgovLyBAdHMtaWdub3JlOiBkZWNvcmF0b3IKQGlubGluZQpmdW5jdGlvbiBoYXNoU3RyKGtleTogc3RyaW5nKTogdTMyIHsKICBpZiAoa2V5ID09PSBudWxsKSB7CiAgICByZXR1cm4gWFhIMzJfU0VFRDsKICB9CgogIHZhciBsZW4gPSBrZXkubGVuZ3RoIDw8IDE7CiAgdmFyIGg6IHUzMiA9IGxlbjsKICBsZXQgaSA9IDA7CgogIGlmIChsZW4gPj0gMTYpIHsKICAgIGxldCBzMSA9IFhYSDMyX1NFRUQgKyBYWEgzMl9QMSArIFhYSDMyX1AyOwogICAgbGV0IHMyID0gWFhIMzJfU0VFRCArIFhYSDMyX1AyOwogICAgbGV0IHMzID0gWFhIMzJfU0VFRDsKICAgIGxldCBzNCA9IFhYSDMyX1NFRUQgLSBYWEgzMl9QMTsKCiAgICAvLyBsZXQgaSA9IDA7CiAgICBsZXQgbiA9IGxlbiAtIDE2OwoKICAgIHdoaWxlIChpIDw9IG4pIHsKICAgICAgczEgPSBtaXgoczEsIGxvYWQ8dTMyPihjaGFuZ2V0eXBlPHVzaXplPihrZXkpICsgaSkpOwogICAgICBzMiA9IG1peChzMiwgbG9hZDx1MzI+KGNoYW5nZXR5cGU8dXNpemU+KGtleSkgKyBpLCA0KSk7CiAgICAgIHMzID0gbWl4KHMzLCBsb2FkPHUzMj4oY2hhbmdldHlwZTx1c2l6ZT4oa2V5KSArIGksIDgpKTsKICAgICAgczQgPSBtaXgoczQsIGxvYWQ8dTMyPihjaGFuZ2V0eXBlPHVzaXplPihrZXkpICsgaSwgMTIpKTsKICAgICAgaSArPSAxNjsKICAgIH0KICAgIGggKz0gcm90bChzMSwgMSkgKyByb3RsKHMyLCA3KSArIHJvdGwoczMsIDEyKSArIHJvdGwoczQsIDE4KTsKICAgIC8vIGxlbiAtPSBpOwogIH0gZWxzZSB7CiAgICBoICs9IFhYSDMyX1NFRUQgKyBYWEgzMl9QNTsKICB9CgogIHZhciBuID0gbGVuIC0gNDsKCiAgd2hpbGUgKGkgPD0gbikgewogICAgaCArPSBsb2FkPHUzMj4oY2hhbmdldHlwZTx1c2l6ZT4oa2V5KSArIGkpICogWFhIMzJfUDM7CiAgICBoICA9IHJvdGwoaCwgMTcpICogWFhIMzJfUDQ7CiAgICBpICs9IDQ7CiAgfQoKICB3aGlsZSAoaSA8IGxlbikgewogICAgaCArPSA8dTMyPmxvYWQ8dTg+KGNoYW5nZXR5cGU8dXNpemU+KGtleSkgKyBpKSAqIFhYSDMyX1A1OwogICAgaCA9IHJvdGwoaCwgMTEpICogWFhIMzJfUDE7CiAgICBpKys7CiAgfQoKICBoIF49IGggPj4gMTU7CiAgaCAqPSBYWEgzMl9QMjsKICBoIF49IGggPj4gMTM7CiAgaCAqPSBYWEgzMl9QMzsKICBoIF49IGggPj4gMTY7CiAgcmV0dXJuIGg7Cn0KCgpleHBvcnQgZnVuY3Rpb24gYmVuY2hPbGQoKTogdTMyIHsKICBsZXQgYWNjOiB1MzIgPSAwOwogIGZvciAobGV0IGkgPSAwOyBpIDwgMTAwMDAwMDsgaSsrKSB7CiAgICBhY2MgKz0gSEFTSDxzdHJpbmc+KFN0cmluZy5mcm9tQ2hhckNvZGUoaSAmIDB4N0YpLnJlcGVhdCgzMikpOwogIH0KICByZXR1cm4gYWNjOwp9CgpleHBvcnQgZnVuY3Rpb24gYmVuY2hOZXcoKTogdTMyIHsKICBsZXQgYWNjOiB1MzIgPSAwOwogIGZvciAobGV0IGkgPSAwOyBpIDwgMTAwMDAwMDsgaSsrKSB7CiAgICBhY2MgKz0gaGFzaFN0cihTdHJpbmcuZnJvbUNoYXJDb2RlKGkgJiAweDdGKS5yZXBlYXQoMzIpKTsKICB9CiAgcmV0dXJuIGFjYzsKfQojIWh0bWwKPHRleHRhcmVhIGlkPSJvdXRwdXQiIHN0eWxlPSJoZWlnaHQ6IDEwMCU7IHdpZHRoOiAxMDAlIiByZWFkb25seT48L3RleHRhcmVhPgo8c2NyaXB0Pgpsb2FkZXIuaW5zdGFudGlhdGUobW9kdWxlX3dhc20sIHsgLyogaW1wb3J0cyAqLyB9KQogIC50aGVuKCh7IGV4cG9ydHMgfSkgPT4gewogICAgY29uc3Qgb3V0cHV0ID0gZG9jdW1lbnQuZ2V0RWxlbWVudEJ5SWQoJ291dHB1dCcpOwogICAgbGV0IHQgPSAwOwoKICAgIHQgPSBwZXJmb3JtYW5jZS5ub3coKTsKICAgIGV4cG9ydHMuYmVuY2hPbGQoKTsKICAgIHQgPSBwZXJmb3JtYW5jZS5ub3coKSAtIHQ7CiAgICBvdXRwdXQudmFsdWUgKz0gJ29sZCBzdHIgaGFzaDogJyArIHQgKyAnIG1zXG4nOwoKICAgIHQgPSBwZXJmb3JtYW5jZS5ub3coKTsKICAgIGV4cG9ydHMuYmVuY2hOZXcoKTsKICAgIHQgPSBwZXJmb3JtYW5jZS5ub3coKSAtIHQ7CiAgICBvdXRwdXQudmFsdWUgKz0gJ25ldyBzdHIgaGFzaDogJyArIHQgKyAnIG1zJzsKICB9KQo8L3NjcmlwdD4=)

[u64 hash tests](https://www.assemblyscript.org/editor.html#IyFvcHRpbWl6ZT1zcGVlZCZydW50aW1lPW5vbmUmbm9Bc3NlcnQKaW1wb3J0IHsgSEFTSCB9IGZyb20gInV0aWwvaGFzaCI7CgpAaW5saW5lIGNvbnN0IFhYSDMyX1AxOiB1MzIgPSAyNjU0NDM1NzYxOwpAaW5saW5lIGNvbnN0IFhYSDMyX1AyOiB1MzIgPSAyMjQ2ODIyNTE5OwpAaW5saW5lIGNvbnN0IFhYSDMyX1AzOiB1MzIgPSAzMjY2NDg5OTE3OwpAaW5saW5lIGNvbnN0IFhYSDMyX1A0OiB1MzIgPSA2NjgyNjUyNjM7CkBpbmxpbmUgY29uc3QgWFhIMzJfUDU6IHUzMiA9IDM3NDc2MTM5MzsKQGlubGluZSBjb25zdCBYWEgzMl9TRUVEOiB1MzIgPSAwOwoKZnVuY3Rpb24gaGFzaDMyKGtleTogdTMyKTogdTMyIHsKICB2YXIgaDogdTMyID0gWFhIMzJfU0VFRCArIFhYSDMyX1A1OwogIGggKz0ga2V5ICogWFhIMzJfUDM7CiAgaCAgPSByb3RsKGgsIDE3KSAqIFhYSDMyX1A0OwogIGggXj0gaCA+PiAxNTsKICBoICo9IFhYSDMyX1AyOwogIGggXj0gaCA+PiAxMzsKICBoICo9IFhYSDMyX1AzOwogIGggXj0gaCA+PiAxNjsKICByZXR1cm4gaDsKfQoKZnVuY3Rpb24gaGFzaDY0KGtleTogdTY0KTogdTMyIHsKICB2YXIgaDogdTMyID0gWFhIMzJfU0VFRCArIFhYSDMyX1A1OwogIGggKz0gPHUzMj5rZXkgKiBYWEgzMl9QMzsKICBoICA9IHJvdGwoaCwgMTcpICogWFhIMzJfUDQ7CiAgaCArPSA8dTMyPihrZXkgPj4gMzIpICogWFhIMzJfUDM7CiAgaCAgPSByb3RsKGgsIDE3KSAqIFhYSDMyX1A0OwogIGggXj0gaCA+PiAxNTsKICBoICo9IFhYSDMyX1AyOwogIGggXj0gaCA+PiAxMzsKICBoICo9IFhYSDMyX1AzOwogIGggXj0gaCA+PiAxNjsKICByZXR1cm4gaDsKfQoKZXhwb3J0IGZ1bmN0aW9uIGJlbmNoT2xkKCk6IHUzMiB7CiAgbGV0IGFjYzogdTMyID0gMDsKICBmb3IgKGxldCBpOiB1NjQgPSAxMDAwMDAwMDA7IGkgPCB1NjQoMTAwMDAwMDAwKSArIDEwMDAwMDAwMDsgaSsrKSB7CiAgICBhY2MgKz0gSEFTSDx1NjQ+KGkpOwogIH0KICByZXR1cm4gYWNjOwp9CgpleHBvcnQgZnVuY3Rpb24gYmVuY2hOZXcoKTogdTMyIHsKICBsZXQgYWNjOiB1MzIgPSAwOwogIGZvciAobGV0IGk6IHU2NCA9IDEwMDAwMDAwMDsgaSA8IHU2NCgxMDAwMDAwMDApICsgMTAwMDAwMDAwOyBpKyspIHsKICAgIGFjYyArPSBoYXNoNjQoaSk7CiAgfQogIHJldHVybiBhY2M7Cn0KIyFodG1sCjx0ZXh0YXJlYSBpZD0ib3V0cHV0IiBzdHlsZT0iaGVpZ2h0OiAxMDAlOyB3aWR0aDogMTAwJSIgcmVhZG9ubHk+PC90ZXh0YXJlYT4KPHNjcmlwdD4KbG9hZGVyLmluc3RhbnRpYXRlKG1vZHVsZV93YXNtLCB7IC8qIGltcG9ydHMgKi8gfSkKICAudGhlbigoeyBleHBvcnRzIH0pID0+IHsKICAgIGNvbnN0IG91dHB1dCA9IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCdvdXRwdXQnKTsKICAgIGxldCB0ID0gMDsKCiAgICB0ID0gcGVyZm9ybWFuY2Uubm93KCk7CiAgICBleHBvcnRzLmJlbmNoT2xkKCk7CiAgICB0ID0gcGVyZm9ybWFuY2Uubm93KCkgLSB0OwogICAgb3V0cHV0LnZhbHVlICs9ICdvbGQgdTY0IGhhc2g6ICcgKyB0ICsgJyBtc1xuJzsKCiAgICB0ID0gcGVyZm9ybWFuY2Uubm93KCk7CiAgICBleHBvcnRzLmJlbmNoTmV3KCk7CiAgICB0ID0gcGVyZm9ybWFuY2Uubm93KCkgLSB0OwogICAgb3V0cHV0LnZhbHVlICs9ICduZXcgdTY0IGhhc2g6ICcgKyB0ICsgJyBtcyc7CiAgfSkKPC9zY3JpcHQ+)

```
old u64 hash (fnv-1a): 281.68000001460314 ms
new u64 hash (xxhash): 189.05499996617436 ms
```

- [x] I've read the contributing guidelines

### CODE REVIEW

File: std/assembly/util/hash.ts [hunk 1]
```diff
@@ -10,63 +8,115 @@ export function HASH<T>(key: T): u32 {
...
+    let s1 = XXH32_SEED + XXH32_P1 + XXH32_P2;
+    let s2 = XXH32_SEED + XXH32_P2;
+    let s3 = XXH32_SEED;
+    let s4 = XXH32_SEED - XXH32_P1;
+
+    let i = 0;
+    len -= 16;
+
+    while (i <= len) {
+      s1 = mix(s1, load<u32>(changetype<usize>(key) + i));
```
Comment (dcodeIO) [inline std/assembly/util/hash.ts:89]: The loads here seem to perform unaligned accesses in 50% of cases, e.g. if len is 18. May make sense to to add an `if` to handle the unaligned 2 bytes before.

Comment (MaxGraey) [reply std/assembly/util/hash.ts:89]: I investigated this again and found couple issues. Fixed

File: std/assembly/util/hash.ts [hunk 2]
```diff
@@ -10,63 +8,110 @@ export function HASH<T>(key: T): u32 {
...
 function hashStr(key: string): u32 {
-  var v = FNV_OFFSET;
-  if (key !== null) {
-    for (let i: usize = 0, k: usize = key.length << 1; i < k; ++i) {
-      v = (v ^ <u32>load<u8>(changetype<usize>(key) + i)) * FNV_PRIME;
+  if (key === null) return XXH32_SEED;
+
+  var len: usize = key.length << 1;
+  var pos = changetype<usize>(key);
+  var h = <u32>len;
```
Comment (dcodeIO) [inline std/assembly/util/hash.ts:76]:
```suggestion
  var h = key.length << 1;
  var len: usize = h;
  var pos = changetype<usize>(key);
```

**Review by dcodeIO — APPROVED** [review]

**Comment by dcodeIO** [comment]
🎉

**Pull request merged**